### PR TITLE
Add default composable templates for new indexing strategy

### LINKED ISSUE #56709: Bundle by default miminal composable templates for the new indexing strategy
As part of the new indexing strategy proposed in this [doc](https://github.com/elastic/kibana/blob/f91c795e30b7221cdce36e235470a4cda82f70b2/docs/ingest_manager/index.asciidoc) we would like to add by default a few v2 templates that will be installed by default.

That will match the following index patterns:

```
logs-*-*
metrics-*-*
events-*-*
```

This is to ensure that the Elastic Agent just works out of the box before Kibana is started. 

The downside of this indices matching the above patterns will potentially have a conflict. One of the ideas was that we provide a config flag in Elasticsearch to disable this setup in 7.x so users can opt out. 

With the flag we didn't need this setting as the v1 template was still honored even if a v2 template was around and only had priority if the flag was set which was the case for data sent from the agent. Now that we don't have the flag anymore I assume we need such a setting or other ideas how to help users that will face this conflict in 7.x.

### PR DESCRIPTION
**Pull request by dakrone**
This commit adds the component and composable templates, as well as ILM policies, for the new
default indexing strategy. It installs:

- logs-default-mappings (component)
- logs-default-settings (component)
- logs-default-policy (ilm policy)
- logs-default-template (composable template)
- metrics-default-mappings (component)
- metrics-default-settings (component)
- metrics-default-policy (ilm policy)
- metrics-default-template (composable template)

These templates and policies are managed by a new x-pack module, `stack`, and can be disabled by
setting `stack.templates.enabled` to `false`.

These ensure that patterns for the `logs-*-*` and `metrics-*-*` indices are set up to create data
streams with the proper mappings and settings.

This also makes changes to the `IndexTemplateRegistry` to support installing component and
composable templates (previously it supported only legacy templates).

Resolves #56709

### CODE REVIEW

**Comment by dakrone** [comment]
@ph @ruflin I requested reviews for you since this uses the templates/policies from Fleet, can you ensure that these are in fact the correct mappings, settings, and policies that we want to use?

**Comment by dakrone** [comment]
@gwbrown pinging you also since you were interested in how I changed `IndexTemplateRegistry` to work for composable and component templates

**Comment by dakrone** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:4]: Could we add some metadata here? It would be nice to indicate in here that these are "stack" templates. The other thing that would be nice is to have the version of the stack that loaded it in here. 

What is the behaviour if the ES is upgraded, will these be overwritten?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:22]: We just renamed these to `dataset.*`: https://github.com/elastic/package-registry/issues/491

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:28]: This is now `name`.

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:36]: I'm torn if we should add these or leave them out.

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:67]: As ECS version is only max about 10 chars long, I wonder if there is a benefit of setting this much lower?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:114]: For `host.*` I was thinking if we should only keep `host.ip`. This ensures that `host` is an object and it is one of the fields which is not a keyword. The other ones should be detected automatically.

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:148]: We have two `_meta` fields?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:17]: Should we have this as `true`? Could this cause problems if a date is a false positive?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:10]: There haven't been any problem with 1024 historically so I'm good with keeping it. But I wonder if there is a value for ES which would preferrable?

File: x-pack/plugin/core/src/main/resources/logs-default-policy.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-policy.json:1]: We should discuss more what our default ILM policy should be moving forward in the context of the new indexing strategy. @jpountz

File: x-pack/plugin/core/src/main/resources/logs-default-settings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-settings.json:8]: Any recommendations here?

File: x-pack/plugin/core/src/main/resources/logs-default-settings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-settings.json:9]: Also looking for recommendations here from the ES team on what to pick.

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:18]: We need to discuss, if dynamic mapping in here should be true or false. Historically I would have said true but seeing that in the future we will also have the possibility to query some fields which are not indexed, I wonder if `false` would be better here?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:151]: Nice!

File: x-pack/plugin/core/src/main/resources/logs-default-settings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-settings.json:13]: Do we still need this? What is the default if we don't set it?

File: x-pack/plugin/core/src/main/resources/logs-default-policy.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-policy.json:12]: Is there a way to add metadata to the ILM policy so we also know which stack version initially loaded it?

File: x-pack/plugin/core/src/main/resources/logs-default-template.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-template.json:3]: Should we set priority to 1 or 10?

**Review by ruflin — COMMENTED** [review]
@dakrone Left quite a few comments on the templates. These should not block the PR itself as we can still follow up making adjustments to the templates.

One thing I was wondering about is on what happens if Elasticsearch is updated and all the assets are already there. Will ES update some of them or leave them as is?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (jpountz) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:17]: I suspect that Lee's motivation is that a value that looks like a date could make dynamic mappings create a date field, which would then fail indexing any document that has a non-date value for this field.

Comment (jpountz) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:17]: should we disable `numeric_detection` too for the same reason?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (jpountz) [inline x-pack/plugin/core/src/main/resources/logs-default-mappings.json:1]: @ruflin would `logs-generic-mappings` be more aligned with Observability in terms of naming?

**Review by jpountz — COMMENTED** [review]
I think we should discuss the implications of making changes to these Elasticsearch default templates. ECS is not set in stone, there are changes that happen from time to time, e.g. a field gets renamed. How would it play with this approach? If the field gets renamed when Elasticsearch upgrades, then wouldn't it fail existing shippers that send documents with the old field name rather than the new one (since our instructions recommend upgrading Elasticsearch first)?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:1]: I think it should be `logs-mappings` and leave out `generic` or `default` as it applies to all datasets.

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:11]: what happens if a `keyword` field mapping is updated to a `wildcard`. Is this a "compatible" mapping change?

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ruflin) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:17]: Can this be enabled / disabled per object or only globally? In general +1 on having it in to indicate if it is true or false. I'm leaning to false. @simitt might have an opinion here.

File: x-pack/plugin/core/src/main/resources/logs-default-settings.json
Comment (urso) [reply x-pack/plugin/core/src/main/resources/logs-default-settings.json:13]: Number of routing shards has been removed in the default Beats templates. We have added it set only for ES 6.x. See 6.8 docs for [Split Index](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/indices-split-index.html#indices-split-index).

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (jpountz) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:11]: you can't replace it on an existing index, but these fields have exactly the same semantics

**Comment by jpountz** [comment]
So @ruflin just explained to me how our packages are going to deal with backward compatibility by having ingest pipelines that will reformat documents that use an old format (e.g. think of a renamed field). Should we introduce an index pipeline here too? It would be initially empty, but this might be important to reserve namespace? Or alternatively should we make the mappings minimal so that we could realistically expect to never have to change the default mappings, e.g. only define a `@timestamp` field?

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
> One thing I was wondering about is on what happens if Elasticsearch is updated and all the assets are already there. Will ES update some of them or leave them as is?

ES uses the version in the template and the version of the `stack` module to determine whether the templates need to be updated, if the `stack` module were updated (for instance, we made changes to the templates), then it replaces them if the existing version is `null` or `< STACK_VERSION`

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:67]: There isn't going to be a benefit really? This is only there to ignore ones where a value was mis-indexed

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:114]: Okay, I will remove the non-ip fields under `host`

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:148]: This is `_meta` on the template itself, not the mapping (which is the other `_meta`). That way when someone retrieves the template they can see where it originated from

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:10]: 1024 is a good default I think

File: x-pack/plugin/core/src/main/resources/logs-default-settings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-settings.json:8]: This is a throughput-storage tradeoff, so the question is just which one we want to default to. This is what the current templates were, but with our other data tiering plans (searchable snapshots), maybe it doesn't make sense any more, I'll add it to the "to discuss" pile

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:18]: I think we should leave dynamic mapping enabled, especially since we're dropping other fields (things like `host.hostname`) that a user may still want to query

File: x-pack/plugin/core/src/main/resources/logs-default-policy.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-policy.json:12]: Not currently, we could potentially add it in the future though

File: x-pack/plugin/core/src/main/resources/logs-default-template.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-template.json:3]: Definitely not 1, because we want ample space for lower-priority templates, 100 seemed reasonable and was confirmed by the core/features team as a reasonable default. I'd prefer to stick with 100

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:11]: I'll collect this as a discussion topic

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:1]: I went with `logs-default-mappings` to allow a user to create their own `logs-mappings` that was explicitly *not* the default one, I also figured it would have a lower chance of existing conflict with any templates a user may have installed

**Comment by dakrone** [comment]
> Should we introduce an index pipeline here too? It would be initially empty, but this might be important to reserve namespace?

I'm not sure we should define the namespace yet, since we don't know whether the pipeline would be generic (both metrics & logs) or specific, so maybe we could leave it until we actually need it?

> Or alternatively should we make the mappings minimal so that we could realistically expect to never have to change the default mappings, e.g. only define a @timestamp field?

Interesting though, I'll add this to the "to discuss" list

@ruflin @jpountz thanks for taking a look, I wrote down a bunch of things to be discussed and I'll put together a document and time to discuss them (and incorporate any other feedback in this PR there)

File: x-pack/plugin/core/src/main/resources/logs-default-mappings.json
Comment (ph) [reply x-pack/plugin/core/src/main/resources/logs-default-mappings.json:36]: @ruflin I would leave it. I think it ecourages good practice to have it.

**Comment by ph** [comment]
@jsvd @acchen97 As discussed in a previous document this is the default templates we would like to ship with ES that follow the indexing strategy document.

**Comment by ruflin** [comment]
I'm wondering if ES should really be responsible to update the template or only to install it the first time. If ES only install it if it is missing, ES can delegate some of the update issues to Ingest Manager.

My thinking at the moment: ES setups up the templates when they don't exist. From their own, they are managed by Ingest Manager and the packages. What this means, that the same templates (or updated ones) are also part of the package-registry. When the Ingest Manager detects the templates changed, it will update them and handle with it all the magic of updating it. But I expect these templates to change very rarely. And if they change, the templates will also be updated in ES for the next minor, so they are installed for new users.

We should keep the fields to a minimum but also not only `@timestamp` to at least enforce some basics of ECS and prevent some historical issues around `host`.

**Comment by dakrone** [comment]
> I'm wondering if ES should really be responsible to update the template or only to install it the first time. If ES only install it if it is missing, ES can delegate some of the update issues to Ingest Manager.

> My thinking at the moment: ES setups up the templates when they don't exist. From their own, they are managed by Ingest Manager and the packages. What this means, that the same templates (or updated ones) are also part of the package-registry. When the Ingest Manager detects the templates changed, it will update them and handle with it all the magic of updating it. But I expect these templates to change very rarely. And if they change, the templates will also be updated in ES for the next minor, so they are installed for new users.

I'm not sure I see what the actual benefits of this approach would be? If possible, I think I'd rather have the same behavior for a user using ES by itself and a user that uses Fleet (ie, they shouldn't have differing mappings and template update behavior)

**Comment by ruflin** [comment]
In the Ingest Manager we already (mostly) implemented all the logic on how to update mappings so Elasticsearch would not have to reimplement the same. 

This is not about Fleet but Ingest Manager, more precisely the Package Manager. I assume any user using Kibana will automatically use it directly or indirectly. The upgrade could even be triggered when Kibana is upgraded or let the user decide when he wants to upgrade the mappings.

Overall I think it is a nice benefit not tying mapping changes to an the upgrade process of the Elasticsearch cluster and let the user decide when he wants to trigger it.

**Comment by albertzaharovits** [comment]
From an Elasticsearch Security perspective, I think it's better that the updates (as well as the initial installation) to policies and templates be handled by ES, not by Fleet or the Ingest Manager.

We currently live with a pressing issue that built-in system users such as `kibana_system`, `remote_monitoring_agent`, and maybe others, which are used by components inside the stack, have high privileges (to install/overwrite templates and policies) just to support the initial setup scenario. After the setup, during ingestion, those high privileges can effectively be dropped, since they're not required anymore.
But we don't have a method to elevate privileges during a setup and we also lack granularity (although we might improve on that soon) when configuring privileges for policies and templates (eg. "can only create policies under the metric* namespace", and "can apply those policies only to indices in the events-* namespace"), there is a single privilege for all policies.

If ES will internally handle stack policies and templates installation and upgrades, we would be relieved that we can lower privileges of built-in system users.

**Comment by dakrone** [comment]
> If ES will internally handle stack policies and templates installation and upgrades, we would be relieved that we can lower privileges of built-in system users.

It would also mean that we wouldn't have to worry about a fleet user that put an ILM policy in place, and then had the user removed/changed, causing the ILM policy not to be able to execute (since policies are executed with the credentials of the last update)

**Comment by ruflin** [comment]
I'm not sure the first problem around user permissions really applies here. Installing assets like templates, ILM policies etc. is the bread and butter of the Package Manager (part of Ingest Manager). We need to do it for all the packages we handle. Even if we don't setup the base templates, we still need all these permissions. Side note: We don't use a system user at the moment.

@dakrone For your comment about ILM policies:

* I don't think we should ever update ILM policies and always assume after we loaded it for the first time, the user modified it. This simplifies the problem.
* What happens in the case a user modifies an ILM policy and then is removed? Does this mean this policy will not be executed anymore?
* How this it work in the case that an the ILM policy was modified with a user logged in via API Key?

File: x-pack/plugin/core/src/main/resources/logs-default-settings.json
Comment (jpountz) [reply x-pack/plugin/core/src/main/resources/logs-default-settings.json:8]: One thing that crossed my mind recently is whether the templates are the right way to configure this default, or whether this would be a node-level default, e.g. we could default to `best_compression` on the warm nodes.

**Comment by dakrone** [comment]
I've made updates to this PR based on the feedback received, thanks for chiming in everyone (please continue to do so)

**Comment by ruflin** [comment]
Getting back on the naming of templates and components. To explain the problem, first our naming scheme. If someone ingests logs for the `nginx.access` dataset, he will have the following templates / components:

* logs-nginx.access (template)
* logs-nginx.access-mappings (component)
* logs-nginx.access-settings (component)

Now the above proposed name for the components is `logs-default-mappings`. Now if a user decides to have a dataset called `default`, the names would conflict. In reality, I hope we never have a dataset called `default` but we can't prevent it.

This is why I'm suggesting to go with the following naming to prevent conflicts:

* logs (template)
* logs-mappings (component)
* logs-settings (component)

The above assumes, that a template and a component can have the same name. Because it is still possible, that user would create a `settings` dataset and have then a `logs-settings` template.

For the ILM policy, I don't expect conflicts but I would follow the same schema so call it `logs`. Adding `policy` to the name feels like repetition as the user is already looking at a policy (btw same applies for having template in the template name). I also like `logs-default` as it is indeed the default policy.

If the above does not work, I'm sure we can also make the current proposal work and just tell users not to use the name `default` for their dataset names.

File: x-pack/plugin/core/src/main/resources/logs-default-policy.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-default-policy.json:4]: is this `min_age` required? What is the effect of it?

**Review by ruflin — COMMENTED** [review]
@dakrone One thing I would like to see is some documentation on our decisions made on what fields to pick and which ones not so that if later some asks around what "frozen" is missing in the ILM policy, we can point him somewhere. It will also make sure if we have future discussions around the options (which we will) we can at least start from a foundation. Not sure what the best place is to document these but one suggestion would be inside the JSON with comments. Even though this will not be stored like this in ES in the end, any person going back to trying to change the ILM policy and templates in the code, will directly see the comments.

File: x-pack/plugin/core/src/main/resources/metrics-default-mappings.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/metrics-default-mappings.json:16]: @exekias @roncohen I was wondering if we should add anything related to detecting numbers to our default template?

Comment (exekias) [reply x-pack/plugin/core/src/main/resources/metrics-default-mappings.json:16]: uhm good question, it's my understating that beats already send metrics correctly encoded as numbers in the JSON request. That would make Elasticsearch use float and integer. Probably this is fine as a default, WDYT?

Comment (ruflin) [reply x-pack/plugin/core/src/main/resources/metrics-default-mappings.json:16]: Beats is sending `1.0` if we set it as float instead of `1`? I must confess, I don't remember. But if that is the case, ++ on having even number detection on. 

@dakrone What are all the option we have here?

File: x-pack/plugin/core/src/main/resources/logs-default-policy.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-default-policy.json:4]: It's not required, and 0 is the default so this can be removed

File: x-pack/plugin/core/src/main/resources/metrics-default-mappings.json
Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/metrics-default-mappings.json:16]: @ruflin all `numeric_detection` does is to make strings that are numbers (ie `"1.3"` instead of `1.3`) be numbers rather than strings. If beats is already sending data in the correct format (numbers instead of strings), I don't think it would buy us anything other than requiring the string to be read on every new field introduced.

**Comment by dakrone** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/resources/metrics-default-mappings.json
Comment (ruflin) [reply x-pack/plugin/core/src/main/resources/metrics-default-mappings.json:16]: ++, sounds like we can ignore it then for now.

**Comment by dakrone** [comment]
> Not sure what the best place is to document these but one suggestion would be inside the JSON with comments.

@ruflin I think in the commit message would be good, as they would then be accessible to any future developer by annotating/blaming the code. Once we settle on the way things look and the PR is approved, I can change the squashed commit to contain an explanation of our decisions.

**Comment by dakrone** [comment]
> For the ILM policy, I don't expect conflicts but I would follow the same schema so call it logs. Adding policy to the name feels like repetition as the user is already looking at a policy (btw same applies for having template in the template name). I also like logs-default as it is indeed the default policy.
> If the above does not work, I'm sure we can also make the current proposal work and just tell users not to use the name default for their dataset names.

I'm fine with the proposed names (`logs`, `logs-mappings`, `logs-settings` for the templates, with `logs` as the ILM policy name). I'll make this change.

**Comment by dakrone** [comment]
> @dakrone For the docs, I was also looking for something I can point non-devs to and update over time when we adapt. Sounds more like a docs part?

I think if we're going to document it then maybe somewhere in the Fleet docs (do we have Fleet docs?) it would make sense to describe these templates and the things they install. Is that more of what you were thinking?

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by ruflin** [comment]
@dakrone I realised based on your comment, that a good place to document this could be the base package on our end. As soon as this lands, I will copy over the existing assets to the base package and there I could add the comments in line.

Lets move this forward, docs is not a blocker here.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutComponentTemplateRequest.java
Comment (andreidan) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutComponentTemplateRequest.java:81]: is this used anywhere?

File: x-pack/plugin/stack/build.gradle
Comment (andreidan) [inline x-pack/plugin/stack/build.gradle:27]: do we need to enable this before merging?

File: x-pack/plugin/stack/qa/rest/src/test/resources/rest-api-spec/test/stack/10_basic.yml
Comment (andreidan) [inline x-pack/plugin/stack/qa/rest/src/test/resources/rest-api-spec/test/stack/10_basic.yml:11]: will these fail when missing? Should we match the status code?

**Review by andreidan — APPROVED** [review]
LGTM, thanks Lee, had a few questions on super minor issues

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutComponentTemplateRequest.java
Comment (dakrone) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutComponentTemplateRequest.java:81]: I think this is a leftover, I'll remove it

File: x-pack/plugin/stack/build.gradle
Comment (dakrone) [reply x-pack/plugin/stack/build.gradle:27]: Yep, I'll enable it

File: x-pack/plugin/stack/qa/rest/src/test/resources/rest-api-spec/test/stack/10_basic.yml
Comment (dakrone) [reply x-pack/plugin/stack/qa/rest/src/test/resources/rest-api-spec/test/stack/10_basic.yml:11]: Yeah, these definitely fail when the policy/template/etc is missing

File: x-pack/plugin/core/src/main/resources/logs-template.json
Comment (ruflin) [inline x-pack/plugin/core/src/main/resources/logs-template.json:3]: @dakrone Why is this set to 100? Shouldn't we set it to 1? @jonathan-buttner FYI

Comment (dakrone) [reply x-pack/plugin/core/src/main/resources/logs-template.json:3]: @ruflin hah, you already asked that in https://github.com/elastic/elasticsearch/pull/57629#discussion_r435415763 :) We need space in order for templates to be allowed at a lower priority than these built-in ones (for example, someone installing a template for `*` at priority 10)

Comment (ruflin) [reply x-pack/plugin/core/src/main/resources/logs-template.json:3]: 🤦

**Review by jtibshirani — COMMENTED** [review]
Apologies for the very late question -- I was wondering why 'mappings' and 'settings' are pulled out into separate components, as opposed to just being defined on the template itself? Do we anticipate that users will reuse these components in their own templates?

Also is there a reason the settings and mappings live in different components? I would guess that you would reuse both the mappings and settings together.

**Comment by dakrone** [comment]
@jtibshirani I pulled them out to follow the behavior that they were currently in for the ingest manager, maybe @ph or @ruflin can comment about why they were separate?

On the other hand, it is nice to be able to substitute your own mappings for logs and still re-use the settings for the bundled templates, for example someone could use: `composed_of: [logs-mappings, my-custom-mappings, logs-settings]` where they have more control over the fine-grained merge order when these are separated. I think users may still want to use the settings (for example, lifecycle policy) without necessarily using all the mappings?

**Comment by ruflin** [comment]
@jtibshirani The use case mentioned by @dakrone is exactly why mappings and settings are separate. We expect users to reused the same settings and we also do it by default. The packages we install use the "gobal" settings by default as a component.

**Pull request merged**
